### PR TITLE
Engine - Implementation Sophos/Firewall decoder

### PR DESCRIPTION
**Pull request by NahuFigueroa97**
|Related issue|
|---|
|#15851|

## Description

With the aim of expanding the coverage of the decoders, this PR adds a decoder for Sophos/Firewall.

## Configuration options

Validate the new decoder:
- ./build/main catalog validate decoder/sophos-ldp/0 < /home/vagrant/engine/wazuh/src/engine/ruleset/decoders/sophos/firewall.yml

Create decoder:
- ./build/main  catalog create decoder < /home/vagrant/engine/wazuh/src/engine/ruleset/decoders/sophos/firewall.yml

Update environment
- ./build/main catalog update environment/wazuh/0 < /home/vagrant/engine/wazuh/src/engine/ruleset/environments/wazuh-environment.yml

Test new decoder
- ./build/main test -q 1

## Logs/Alerts example

- <30>device="SFW" date=2020-05-18 time=14:38:37 timezone="CEST" device_name="XG230" device_id=1234567890123456 log_id=010101600001 log_type="Firewall" log_component="Firewall Rule" log_subtype="Allowed" status="Allow" priority=Information duration=11 fw_rule_id=21 policy_type=1 user_name="" user_gp="" iap=0 ips_policy_id=0 appfilter_policy_id=0 application="HTTP" application_risk=1 application_technology="Browser Based" application_category="General Internet" in_interface="Port1" out_interface="Port2" src_mac=00:00:00:00:00:00 src_ip=1.128.3.4 src_country_code=R1 dst_ip=91.228.167.86 dst_country_code=SVK protocol="TCP" src_port=62841 dst_port=80 sent_pkts=6  recv_pkts=5 sent_bytes=459 recv_bytes=606 tran_src_ip=213.167.51.66 tran_src_port=0 tran_dst_ip="" tran_dst_port=0 srczonetype="LAN" srczone="LAN" dstzonetype="WAN" dstzone="WAN" dir_disp="" connevent="Stop" connid="1617925280" vconnid="" hb_health="No Heartbeat" message="" appresolvedby="Signature" app_is_cloud=0

## Tests

Event:
- <30>device="SFW" date=2020-05-18 time=14:38:37 timezone="CEST" device_name="XG230" device_id=1234567890123456 log_id=010101600001 log_type="Firewall" log_component="Firewall Rule" log_subtype="Allowed" status="Allow" priority=Information duration=11 fw_rule_id=21 policy_type=1 user_name="" user_gp="" iap=0 ips_policy_id=0 appfilter_policy_id=0 application="HTTP" application_risk=1 application_technology="Browser Based" application_category="General Internet" in_interface="Port1" out_interface="Port2" src_mac=00:00:00:00:00:00 src_ip=1.128.3.4 src_country_code=R1 dst_ip=91.228.167.86 dst_country_code=SVK protocol="TCP" src_port=62841 dst_port=80 sent_pkts=6  recv_pkts=5 sent_bytes=459 recv_bytes=606 tran_src_ip=213.167.51.66 tran_src_port=0 tran_dst_ip="" tran_dst_port=0 srczonetype="LAN" srczone="LAN" dstzonetype="WAN" dstzone="WAN" dir_disp="" connevent="Stop" connid="1617925280" vconnid="" hb_health="No Heartbeat" message="" appresolvedby="Signature" app_is_cloud=0

Output:
```
{
    "wazuh": {
        "queue": 49,
        "origin": "/dev/stdin"
    },
    "event": {
        "original": "<30>device=\"SFW\" date=2020-05-18 time=14:38:37 timezone=\"CEST\" device_name=\"XG230\" device_id=1234567890123456 log_id=010101600001 log_type=\"Firewall\" log_component=\"Firewall Rule\" log_subtype=\"Allowed\" status=\"Allow\" priority=Information duration=11 fw_rule_id=21 policy_type=1 user_name=\"\" user_gp=\"\" iap=0 ips_policy_id=0 appfilter_policy_id=0 application=\"HTTP\" application_risk=1 application_technology=\"Browser Based\" application_category=\"General Internet\" in_interface=\"Port1\" out_interface=\"Port2\" src_mac=00:00:00:00:00:00 src_ip=1.128.3.4 src_country_code=R1 dst_ip=91.228.167.86 dst_country_code=SVK protocol=\"TCP\" src_port=62841 dst_port=80 sent_pkts=6  recv_pkts=5 sent_bytes=459 recv_bytes=606 tran_src_ip=213.167.51.66 tran_src_port=0 tran_dst_ip=\"\" tran_dst_port=0 srczonetype=\"LAN\" srczone=\"LAN\" dstzonetype=\"WAN\" dstzone=\"WAN\" dir_disp=\"\" connevent=\"Stop\" connid=\"1617925280\" vconnid=\"\" hb_health=\"No Heartbeat\" message=\"\" appresolvedby=\"Signature\" app_is_cloud=0",
        "action": "allowed",
        "code": 10101600001,
        "dataset": "sophos.xg",
        "duration": 11,
        "end": "2020-05-18T14:38:37",
        "kind": "event",
        "module": "sophos",
        "outcome": "success",
        "severity": 6,
        "start": "2020-05-18T14:38:37",
        "timezone": "CEST",
        "type": [
            "allowed",
            "connection",
            "start"
        ]
    },
    "destination": {
        "bytes": 606,
        "ip": "91.228.167.86",
        "nat": {
            "ip": "\" tran_dst_port=0 srczonetype=\"LAN"
        },
        "port": 80
    },
    "fileset": {
        "name": "xg"
    },
    "input": {
        "type": "log"
    },
    "log": {
        "level": "Information"
    },
    "network": {
        "transport": "tcp"
    },
    "observer": {
        "egress": {
            "interface": {
                "name": "Port2"
            },
            "zone": "WAN"
        },
        "ingress": {
            "interface": {
                "name": "Port1"
            },
            "zone": "LAN"
        },
        "product": "XG",
        "serial_number": 1234567890123456,
        "type": "firewall",
        "vendor": "Sophos"
    },
    "related": {
        "ip": [
            "1.128.3.4",
            "91.228.167.86",
            "213.167.51.66",
            "\" tran_dst_port=0 srczonetype=\"LAN"
        ],
        "user": [
            "\" user_gp=\"\" iap=0 ips_policy_id=0 appfilter_policy_id=0 application=\"HTTP"
        ]
    },
    "rule": {
        "id": 21,
        "ruleset": 1
    },
    "service": {
        "type": "sophos"
    },
    "sophos": {
        "xg": {
            "app_is_cloud": 0,
            "application_risk": 1,
            "connid": 1617925280,
            "device": "SFW",
            "device_name": "XG230",
            "dst_country_code": "SVK",
            "dst_zone_type": "WAN",
            "log_component": "Firewall Rule",
            "log_id": 10101600001,
            "log_subtype": "Allowed",
            "log_type": "Firewall",
            "priority": "Information",
            "src_country_code": "R1",
            "status": "Allow"
        }
    },
    "source": {
        "bytes": 459,
        "ip": "1.128.3.4",
        "mac": "00:00:00:00:00:00",
        "nat": {
            "ip": "213.167.51.66",
            "port": 0
        },
        "packets": 6,
        "port": 62841,
        "user": {
            "name": "\" user_gp=\"\" iap=0 ips_policy_id=0 appfilter_policy_id=0 application=\"HTTP"
        }
    },
    "tags": [
        "forwarded",
        "preserve_original_event",
        "sophos-xg"
    ],
    "\\@timestamp": "2020-05-18T14:38:37"
}
```

### CODE REVIEW

File: src/engine/ruleset/decoders/sophos/firewall.yml [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    map:
+      - event.severity: 4
+  - check:
+      - ~tmp.priority: Notification
+    map:
+      - event.severity: 5
+  - check:
+      - ~tmp.priority: Information
+    map:
+      - event.severity: 6
```
Comment (JorgeMarinoDev) [inline src/engine/ruleset/decoders/sophos/firewall.yml:73]: Please add an empty newline after each check-map section to improve readability of the conditions.

**Review by JorgeMarinoDev — CHANGES_REQUESTED** [review]
Good job here.
Based on already approved decoders I noticed it's a good idea to improve readability in these cases where lots of conditions are checked consecutively.

File: src/engine/ruleset/decoders/sophos/firewall.yml [hunk 2]
```diff
@@ -0,0 +1,129 @@
+---
+name: decoder/sophos-firewall/0
+
+metadata:
+  module: sophos
+  title: Sophos logs decoder
+  description: Decoder for sophos logs
+  author:
```
Comment (soynof) [inline src/engine/ruleset/decoders/sophos/firewall.yml:8]: Improve (documentation): add compatibility version

File: src/engine/ruleset/decoders/sophos/firewall.yml [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+sources:
+  - decoder/queue-syslog/0
+  - decoder/queue-localfile/0
+
+parse:
+  logpar:
+    # <30>device="SFW" date=2020-05-18 time=14:38:45 timezone="CEST" device_name="XG230" device_id=1234567890123456 log_id=010101600001 log_type="Firewall" log_component="Firewall Rule" log_subtype="Allowed" status="Allow" priority=Information duration=0 fw_rule_id=61 policy_type=2 user_name="elastic@user.local" user_gp="elastic.group.local" iap=0 ips_policy_id=11 appfilter_policy_id=0 application="" application_risk=0 application_technology="" application_category="" in_interface="ipsec0" out_interface="Port2" src_mac=00:00:00:00:00:00 src_ip=10.84.234.7 src_country_code=R1 dst_ip=172.16.34.50 dst_country_code=R1 protocol="TCP" src_port=58543 dst_port=443 sent_pkts=0  recv_pkts=0 sent_bytes=0 recv_bytes=0 tran_src_ip="" tran_src_port=0 tran_dst_ip="" tran_dst_port=0 srczonetype="VPN" srczone="VPN" dstzonetype="VPN" dstzone="VPN" dir_disp="" connevent="Start" connid="1615935064" vconnid="" hb_health="No Heartbeat" message="" appresolvedby="Signature" app_is_cloud=0
+    - event.original: \<<~log.head_message>><~log.payload_messge>
```
Comment (soynof) [inline src/engine/ruleset/decoders/sophos/firewall.yml:24]: Improve (design): if it's possible, use only one temporal variable like ~tmp

**Pull request merged**